teuthology/task/kernel.py

```diff
@@ -910,16 +910,14 @@ def grub2_kernel_select_generic(remote, newversion, ostype):
     entry_num = 0
     if '\nmenuitem ' not in grub2conf:
         # okay, do the newer (el8) grub2 thing
-        proc = remote.run(args=['sudo', 'bash', '-c',
-                                'cat /boot/loader/entries/*.conf'],
+        proc = remote.run(args=['/bin/ls', '/boot/loader/entries'],
```

Comment:
Please use remote.sh()

Reply:
@kshtsk dont you think its fine to use either remote.run or remote.sh?, just trying to understand (.sh seems to be shortcut method for run and it returnds stdout by default compared to remote object returned by run method)

Reply:
@vasukulkarni no, I've already cleaned up the kernel task from the cStringIO usage with a purpose of switching to py3, if this will be merged, I will have to fix it again in order py3 tests are passing for py3, I can't merge it right now because this branch is not tested yet, while it can run some of py27 based runs.

Reply:
@kshtsk maybe you missed my question, I am trying to figure out what is the difference between using remote.run and remote.sh method, they have different purpose right? one returns remote object other is a shortcut for the same .run method but returns stdout,  I am not understanding where py3 is playing role?

Reply:
okay, it is the same but two things: sh always return an 'str' obj, so you do not need provide stdout parameter with a stream obj., and first positional argument is the same as 'args' in run. But yes, they are equivalent.

Reply:
I think I understand your concern now, you commented on wrong remote.run, you should have commented on remote.run that uses StringIO(), thats the one needing change, not this remote.run.

Reply:
Just for the note: both 'run' and 'sh' supports list and str for args, so the next calls are kinda identical:
```
remote.run(args=['/bin/ls', '/boot/loader/entries'])
remote.run(args='/bin/ls /boot/loader/entries')
remote.sh(['/bin/ls', '/boot/loader/entries'])
remote.sh('/bin/ls /boot/loader/entries')
```
The main advantage of the run at the moment is that it can be post-process proc object of runs executed in background, but I'm thing of improving the sh so you can give an object reference or a callback function. The idea is to avoid bootstrapping, it is good for code readability.

Reply:
> I think I understand your concern now, you commented on wrong remote.run, you should have commented on remote.run that uses StringIO(), thats the one needing change, not this remote.run.

I commented under 913 line, if you take a look at 914 you'll see StringIO :-D

Reply:
@kshtsk can you merge these py3 fixes?  i'm just copying and modifying the adjacent code

Reply:
I just read bit of kernel task code now, I see its mix of remote.run and .sh, .run has kind of long legacy,  the best thing is to leave it to user to decide based on return values, .run still has advantages where you want to use remote object in the test code.

Reply:
@kshtsk I'd suggest opening a pr with teh changes on a per-task basis so we can merge and work on top of that.  Right now it's a lose-lose because I can't see what the target interface/methods are that I should use, and even if I did, it would still be creating merge conflicts for you and require a rebase if your changes go in after mine.

Reply:
> @kshtsk I'd suggest opening a pr with teh changes on a per-task basis so we can merge and work on top of that. Right now it's a lose-lose because I can't see what the target interface/methods are that I should use, and even if I did, it would still be creating merge conflicts for you and require a rebase if your changes go in after mine.

This is exactly what is going on, most of the code is merged already. I do my best of fixing task. But the main code of run functions is not well tested yet. Some of the jobs can fails. Corresponding PR is #1362.

For this one change, there will no be merge conflict, we are discussing about your code changes, this is why a do a review of the code to avoid merge conflict, and it is possible to fix that before merge.